sdk/src/azure/iot/az_iot_provisioning_client.c

```diff
@@ -270,46 +268,69 @@ AZ_INLINE az_result _az_iot_provisioning_client_parse_payload_error_code(
 
 AZ_INLINE az_result _az_iot_provisioning_client_payload_registration_result_parse(
     az_json_parser* jp,
-    az_json_token_member* tm,
     az_iot_provisioning_client_registration_result* out_state)
 {
-  if (tm->token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
+  if (jp->token.kind != AZ_JSON_TOKEN_BEGIN_OBJECT)
   {
     return AZ_ERROR_PARSER_UNEXPECTED_CHAR;
   }
 
   bool found_assigned_hub = false;
   bool found_device_id = false;
 
-  while ((!(found_device_id && found_assigned_hub))
-         && az_succeeded(az_json_parser_parse_token_member(jp, tm)))
+  AZ_RETURN_IF_FAILED(az_json_parser_next_token(jp));
+
+  while (jp->token.kind != AZ_JSON_TOKEN_END_OBJECT)
   {
-    if (az_span_is_content_equal(AZ_SPAN_FROM_STR("assignedHub"), tm->name))
+    if (az_json_token_is_text_equal(&jp->token, AZ_SPAN_FROM_STR("assignedHub")))
     {
+      AZ_RETURN_IF_FAILED(az_json_parser_next_token(jp));
+      if (jp->token.kind != AZ_JSON_TOKEN_STRING)
+      {
+        return AZ_ERROR_ITEM_NOT_FOUND;
```

Comment:
> Was this the error that the previous function was returning?

Yes, I didn't change that. The call to `az_json_token_get_string` used to fail with `AZ_ERROR_ITEM_NOT_FOUND` if token type wasn't string.

https://github.com/Azure/azure-sdk-for-c/blob/f31482c03a12f462bea8590f56bd0dd4d1b02717/sdk/src/azure/core/az_json_token.c#L23-L31

> I think "not found" is less informative than something like "malformed JSON" which would be the case here when the key exists but the string value doesn't.

Makes sense. Let me know if you still want me to update the `az_result` for this case specifically in `_az_iot_provisioning_client_payload_registration_result_parse`, or leave it as is,

Keep in mind though, it could also be that the key exists, but instead of it being a string, it got changed to a JSON number or boolean true/false, or `null` instead.

> Token not found could be interpreted as a protocol change bug

So, what `az_result` are you thinking that should translate to?